src/librustc_passes/intrinsicck.rs

```diff
@@ -119,6 +122,262 @@ impl ExprVisitor<'tcx> {
+                        Some(match self.tcx.sess.target.ptr_width {
+                            16 => InlineAsmType::VecI16(fields.len() as u64),
+                            32 => InlineAsmType::VecI32(fields.len() as u64),
+                            64 => InlineAsmType::VecI64(fields.len() as u64),
+                            _ => unreachable!(),
+                        })
+                    }
+                    ty::Float(FloatTy::F32) => Some(InlineAsmType::VecF32(fields.len() as u64)),
+                    ty::Float(FloatTy::F64) => Some(InlineAsmType::VecF64(fields.len() as u64)),
+                    _ => None,
+                }
+            }
+            _ => None,
+        };
+        let asm_ty = match asm_ty {
+            Some(asm_ty) => asm_ty,
+            None => {
+                let msg = &format!("cannot use value of type `{}` for inline assembly", ty);
+                let mut err = self.tcx.sess.struct_span_err(expr.span, msg);
+                err.note(
+                    "only integers, floats, SIMD vectors, pointers and function pointers \
+                     can be used as arguments for inline assembly",
+                );
+                err.emit();
+                return None;
+            }
+        };
+
+        // Check that the type implements Copy. The only case where this can
+        // possibly fail is for SIMD types which don't #[derive(Copy)].
+        if !ty.is_copy_modulo_regions(self.tcx, self.param_env, DUMMY_SP) {
+            let msg = "arguments for inline assembly must be copyable";
+            let mut err = self.tcx.sess.struct_span_err(expr.span, msg);
+            err.note(&format!("`{}` does not implement the Copy trait", ty));
+            err.emit();
+        }
+
+        // Ideally we wouldn't need to do this, but LLVM's register allocator
+        // really doesn't like it when tied operands have different types.
+        //
+        // This is purely an LLVM limitation, but we have to live with it since
+        // there is no way to hide this with implicit conversions.
+        //
+        // For the purposes of this check we only look at the `InlineAsmType`,
+        // which means that pointers and integers are treated as identical (modulo
+        // size).
+        if let Some((in_expr, Some(in_asm_ty))) = tied_input {
+            if in_asm_ty != asm_ty {
+                let msg = &format!("incompatible types for asm inout argument");
+                let mut err = self.tcx.sess.struct_span_err(vec![in_expr.span, expr.span], msg);
+                err.span_label(
+                    in_expr.span,
+                    &format!("type `{}`", self.tables.expr_ty_adjusted(in_expr)),
+                );
+                err.span_label(expr.span, &format!("type `{}`", ty));
+                err.note("asm inout arguments must have the same type");
+                err.note("unless they are both pointers or integers of the same size");
```

Comment:
These two lines constitute effectively a single note, and so should be reworded so that it fits into one `.note()` call.